Add python only default init expression; Implement stft, hann/hamming/bartlett window.

ssnl opened the pull request:
This PR

1. Implements short-time fourier transform #3775 with tests against Scipy.
2. Adds `atan2` for cuda double & half tensors. Refactored `atan2` to `THCNumerics`, Op to `THCTensorMathPointwise.cuh`.
3. Adds support for python only default init in the parsers. This allows us to write argument with a default value that can either cause ambiguity in c++ (e.g., `Scalar p` in `norm`) or have a type that doesn't allow default value `None/NULL/nullptr` (e.g., `int64_t fft_size` in `stft`).
  a. Changes `norm` to also use this in both `Declarations.cwrap` for #1419 . When `Tensor` and `Variable` merge, it should fix the issues in #1419 .
4. Implements three commonly used window functions (Hann, Hamming, Bartlett) with tests against Scipy

On 3: Now in `python_variable_methods.cpp`, if an argument has python only default value, it will be generated like the following. I added the comment for clarification purpose. It won't be in actual code. 
```
static PyObject * THPVariable_stft(PyObject* self, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "stft(int64_t frame_length, int64_t hop, int64_t fft_size=None, Tensor window=None, int64_t pad_end=0)",
  });
  auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
  PyObject* parsed_args[6];
  auto r = parser.parse(args, kwargs, parsed_args);
  if (r.idx == 0) {
    Tensor & self = self_;
    int64_t frame_length = r.toInt64(0);
    int64_t hop = r.toInt64(1);
    const Tensor & window = r.tensor(3);
    int64_t pad_end = r.toInt64(4);
    // fft_size has python only default, so it will be generated last
    int64_t fft_size = r.toInt64WithDefault(2, frame_length);  
    return wrap(dispatch_stft(self, frame_length, hop, fft_size, window, pad_end));
  }
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}
```

ssnl committed:
implement stft

fmassa commented on aten/src/THC/generic/THCTensorMathPointwise.cu:
```diff
@@ -68,6 +68,42 @@ IMPLEMENT_CUDA_TENSOR_BASIC_FUNC(  abs, THCNumerics<real>::abs,   Real)
+
+struct TensorATan2Op {
+  __device__ __forceinline__ void operator()(float* out, float* a, float* b) {
+    *out = atan2f(*a, *b);
```
Nit: wouldn't it be better to add these to THCNumerics? And eventually add support for half as well?

colesbury commented on aten/src/ATen/native/spectral_ops.cpp:
```diff
@@ -0,0 +1,89 @@
+  // k[omega, t] = cos (2 pi omega t / N) - j sin (2 pi omega t / N)
+  double N = static_cast<double>(frame_length);
+  auto arange = self.type().arange(0, frame_length).unsqueeze_(1);
+  auto arange_2d = arange.mm(arange.t()).mul_(M_PI * 2. / N);
```
You can also use [`at::ger`](http://pytorch.org/docs/0.3.0/torch.html#torch.ger) to compute outer products.

I'm a little confused by the scaling. `arange_2d` ends up being proportional to `N` because the multiplication by `1 / N` happens after the outer product. Is that correct?

colesbury commented on test/test_torch.py:
```diff
@@ -2561,6 +2561,71 @@ def test(M):
         self._test_det(self, lambda x: x)
 
+    @staticmethod
+    def _test_stft(self, conv_fn):
```
I would be more confident in these tests if you added a few known sequences and check their stft. Things like constants, +1/-1/+1/-1/..., and maybe a few other simple signals.

colesbury commented on torch/functional.py:
```diff
@@ -263,3 +263,46 @@ def det(var):
+
+
+def stft(var, frame_length, hop, window=None, pad_end=0):
+    r"""stft(signal, frame_length, hop, window=None, pad_end=0) -> Tensor, Tensor
```
Generally, you don't rewrite the signature for Python functions. https://www.python.org/dev/peps/pep-0257/

ssnl committed:
addressed comments; implemented window functions; added support for python only default initialization

SsnL commented:
In the new commit are:
1. Added test for `stft` against Scipy
2. Moved the `atan2` Op to `THCNumerics`
3. Added support for python only default init in the parsers. This allows us to wraite arguments with default values that either causes ambiguity in c++ (e.g., `norm`) or has a type that doesn't allow default value `nullptr` (e.g., `fft_size` in `stft`).
4. Added three commonly used window functions. 

On 3: Now in `python_variable_methods.cpp`, if an argument has python only default value, it will be generated like the following: 
```
static PyObject * THPVariable_stft(PyObject* self, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "stft(int64_t frame_length, int64_t hop, int64_t fft_size=None, Tensor window=None, int64_t pad_end=0)",
  });
  auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
  PyObject* parsed_args[6];
  auto r = parser.parse(args, kwargs, parsed_args);
  if (r.idx == 0) {
    Tensor & self = self_;
    int64_t frame_length = r.toInt64(0);
    int64_t hop = r.toInt64(1);
    const Tensor & window = r.tensor(3);
    int64_t pad_end = r.toInt64(4);
    // fft_size has python only default, so it will be generated last
    int64_t fft_size = r.toInt64WithDefault(2, frame_length);  
    return wrap(dispatch_stft(self, frame_length, hop, fft_size, window, pad_end));
  }
  Py_RETURN_NONE;
  END_HANDLE_TH_ERRORS
}
```

TODO: add tests for window functions

fmassa commented on torch/functional.py:
```diff
@@ -263,3 +265,166 @@ def det(var):
+        raise ValueError('window_length must be positive')
+    return hamming_window(window_length, periordic=periordic, alpha=0.5, beta=0.5)
+
+def hamming_window(window_length, periordic=True, alpha=0.53836, beta=0.46164):
```
nit: did you mean `periodic`?

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

soumith commented:
@SsnL once you resolve merge conflict, this is good to go.

SsnL commented:
@soumith merged & tested :) thanks!

faroit commented:
@SsnL I think it would be nice to also have an inverse stft

SsnL commented:
@faroit Yes, it's on my list of things to do. Thanks for bringing it up.

SsnL commented:
@faroit thanks!
